project4/3700crawler

```diff
@@ -134,21 +174,56 @@ class WebCrawler:
             exit(1)
         self.session = match.group(1)
 
-        self.set_cookie()
-
-        login_resp = self.POST(LOGIN_FORM, make_formstr({
+        login_resp = self.POST(LOGIN_FORM, LOGIN_FORM, make_formstr({
             'username': self.username,
             'password': self.password,
             'csrfmiddlewaretoken': self.csrf,
             'submit': 'Log in',
-            'next': '/fakebook/'
+            'next': r'%2Ffakebook%2F'
         }))
         print(login_resp)
+        match = re.search('[Ss]et-[Cc]ookie:\W*sessionid=(\w+);', login_resp)
+        if (not match):
+            print('Couldn\'t extract SessionID token, aborting')
+            print(login_resp)
+            exit(1)
+        self.session = match.group(1)
+
+        return self.traverse_to(get_location(login_resp))
+
+def get_linkset(origin, page):
+    matches = re.findall('href=[\'"]([/\w]+)', page)
+    for match in matches:
+        print(match)
+    return set()
 
+def hunt_key(page):
+    return None
 
 def crawl(crawler):
     """Runs the web crawler"""
-    crawler.login()
+    visited = set()
+    unvisited = set()
+    secret_keys = []
+
+    login_page = crawler.login()
+    print(login_page)
+    unvisited = get_linkset(ROOT, login_page)
+
+    while len(unvisited) is not 0:
+        url = unvisited.pop()
+        visited.add(url)
+
+        page = crawler.traverse_to(url)
+        if page is None:
+            continue
+        secret_key = hunt_key(page)
+        if secret_key:
+            secret_keys.append(secret_key)
+
+        unvisited |= (get_linkset(ROOT, page) - visited)
```

Comment:
Can you explain what this line does?

Reply:
Thsat adds to the unvisited set all links the set of all links that were in the page less the set of pages aready visited

Reply:
Ah, okay. I was just unsure about the `|=` (haven't started using Python until this class).